mixnet/client/src/sender.rs

```diff
@@ -109,13 +142,129 @@ impl<R: Rng> Sender<R> {
+        let packet_id = PacketId::from_bytes(&bytes);
+        match Body::write_sphinx_packet_bytes(&mut *socket, &bytes).await {
+            Ok(_) => {
+                tracing::info!("Sent a Sphinx packet successuflly to the node: {addr}");
+                Self::insert_packet_id(&ack_cache, addr, packet_id).await;
 
-        tracing::debug!("Sent a Sphinx packet successuflly to the node: {addr:?}");
+                tokio::spawn(async move {
+                    Self::retry_backoff(
+                        ack_cache,
+                        max_retries,
+                        retry_delay,
+                        packet_id,
+                        bytes,
+                        addr,
+                        arc_socket,
+                    )
+                    .await;
+                });
+                Ok(())
+            }
+            Err(e) => {
+                if let Some(err) = e.downcast_ref::<std::io::Error>() {
+                    match err.kind() {
+                        ErrorKind::BrokenPipe
+                        | ErrorKind::NotConnected
+                        | ErrorKind::ConnectionAborted => {
+                            tracing::warn!("broken pipe error while sending a Sphinx packet to the node: {addr}, try to update the connection and retry");
+                            // update the connection
+                            let mut tcp = TcpStream::connect(addr).await?;
+                            // resend packet immediately
+                            Body::write_sphinx_packet_bytes(&mut tcp, &bytes).await?;
+                            *socket = tcp;
+                            tracing::info!("Sent a Sphinx packet successuflly to the node: {addr}");
+                            Self::insert_packet_id(&ack_cache, addr, packet_id).await;
+                            tokio::spawn(async move {
+                                Self::retry_backoff(
+                                    ack_cache,
+                                    max_retries,
+                                    retry_delay,
+                                    packet_id,
+                                    bytes,
+                                    addr,
+                                    arc_socket,
+                                )
+                                .await;
+                            });
+                            Ok(())
+                        }
+                        _ => Err(e),
+                    }
+                } else {
+                    Err(e)
+                }
+            }
+        }
+    }
 
-        Ok(())
+    async fn insert_packet_id(
+        ack_cache: &Arc<Mutex<HashMap<SocketAddr, HashSet<PacketId>>>>,
+        addr: SocketAddr,
+        packet_id: PacketId,
+    ) {
+        ack_cache
+            .lock()
+            .await
+            .entry(addr)
+            .or_insert_with(HashSet::new)
+            .insert(packet_id);
+    }
+
+    async fn retry_backoff(
```

Comment:
This method is way too big and nested. Couldn't it be splited in different functions?